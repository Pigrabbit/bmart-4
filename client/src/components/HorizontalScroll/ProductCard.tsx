import React from 'react'
import styled from 'styled-components'
import { parseToLocalMoneyString } from '../../utils/parser'
import { ProductType } from './HorizontalScroll'
import { STYLES } from '../../utils/styleConstants'

type Props = {
  product: ProductType
}

const StyledContainer = styled.div`
  width: 120px;
  flex: 0 0 auto;
  margin-right: ${STYLES.margin};
`
const StyledThumbnail = styled.div`
  img {
    width: 100%;
    height: 100%;
  }
`
const StyledContent = styled.div`
  line-height: 20px;
  .price {
    font-weight: 700;
  }
`

export const ProductCard = (props: Props) => {
  const { price, name, thumbnail } = props.product

  return (
    <StyledContainer>
      <StyledThumbnail>
        <img src={thumbnail} alt="" />
      </StyledThumbnail>
      <StyledContent>
        <div>{name}</div>
        <div className="price">{parseToLocalMoneyString(price)}Ïõê</div>
      </StyledContent>
    </StyledContainer>
  )
}
