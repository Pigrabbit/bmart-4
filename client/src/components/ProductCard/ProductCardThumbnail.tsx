import React from 'react'
import styled from 'styled-components'
import { COLORS } from '../../utils/styleConstants'

type Props = {
  isLiked: boolean
  lazyLoad?: boolean
  isSoldOut: boolean
  thumbnailSrc: string
  toggleProductLike: (e: React.MouseEvent) => void
  seperateClickEventHandler: (e: React.MouseEvent) => void
}

const StyledThumbnail = styled.div`
  position: relative;
  font-size: 0;
  .icon-wrap {
    font-size: 24px;
    line-height: 24px;
    width: 24px;
    height: 24px;
    position: absolute;
    right: 4px;
    bottom: 4px;
    z-index: 10;
    .icon {
      color: #eee;
      border-radius: 50%;
      &.like {
        color: ${COLORS.red};
      }
    }
  }
  .thumbnail-wrap {
    width: 100%;
    height: 100%;
    border-radius: 6px;
    filter: brightness(0.96);
    overflow: hidden;
    position: relative;
    &.alt::before {
      content: '';
      display: block;
      padding: 50%;
      background-color: #fafafa;
    }
    &.sold-out::before {
      content: 'sold out';
      display: block;
      font-size: 24px;
      color: white;
      font-family: 'BMHANNAPro';
      position: absolute;
      background-color: rgba(0, 0, 0, 0.4);
      left: 0;
      top: 0;
      right: 0;
      bottom: 0;
      z-index: 10;
      display: flex;
      justify-content: center;
      align-items: center;
    }
    .thumbnail {
      width: 100%;
      height: 100%;
      position: relative;
    }
  }
`

export const ProductCardThumbnail = (props: Props) => {
  const { thumbnailSrc, lazyLoad, isLiked, isSoldOut } = props

  const renderThumbnail = () => {
    if (lazyLoad === undefined) {
      return (
        <div className={`thumbnail-wrap ${isSoldOut ? 'sold-out' : ''}`}>
          <img className="thumbnail" src={thumbnailSrc} alt="" />
        </div>
      )
    } else {
      if (!lazyLoad) {
        return (
          <div className="thumbnail-wrap alt">
            <img className="thumbnail" src={''} data-lazy={thumbnailSrc} alt="" />
          </div>
        )
      } else {
        return (
          <div className={`thumbnail-wrap ${isSoldOut ? 'sold-out' : ''}`}>
            <img className="thumbnail" src={thumbnailSrc} data-lazy={thumbnailSrc} alt="" />
          </div>
        )
      }
    }
  }

  return (
    <StyledThumbnail onDoubleClick={props.seperateClickEventHandler}>
      {(lazyLoad === undefined || lazyLoad) && (
        <div className="icon-wrap" onClick={props.toggleProductLike}>
          {isLiked ? (
            <i className="icon like">heart_circle</i>
          ) : (
            <i className="icon">heart_circle_fill</i>
          )}
        </div>
      )}
      {renderThumbnail()}
    </StyledThumbnail>
  )
}
