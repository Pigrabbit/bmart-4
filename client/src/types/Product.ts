export type ProductType = {
    id: number
    name: string
    coupang_product_id: number
    category: string
    price: number
    base_price?: number
    discount_rate?: number
    thumbnail_src: string
    created_at: string
    stock_count: number
    sold_count: number
    description?: string
}
