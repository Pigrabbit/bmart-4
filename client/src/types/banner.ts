export type BannerType = { src: string; href: string }
