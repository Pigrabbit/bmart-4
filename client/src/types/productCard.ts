export type ProductCardType = {
  id: string
  price: number
  name: string
  thumbnailSrc: string
  coupangProductId: string
  basePrice: number
  discountRate: number
  isLiked: boolean
  stockCount: number
}
