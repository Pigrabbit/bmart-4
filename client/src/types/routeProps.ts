export type ProductDetailRouteProps = {
  // url parameter로 전달되면 string으로 전달되기 때문에 number로 변경해 주어야 함
  productId: string
}
export type CategoryDashboardRouteProps = {
  categoryId: string
}
