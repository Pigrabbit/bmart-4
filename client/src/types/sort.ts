export type SortType = '' | 'priceAsc' | 'priceDesc'
