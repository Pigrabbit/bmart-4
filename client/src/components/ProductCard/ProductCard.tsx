import React, { useState } from 'react'
import styled from 'styled-components'
import { parseToLocalMoneyString } from '../../utils/parser'
import { ProductCardType } from '../../types/productCard'
import { useMutation } from '@apollo/client'
import { useHistory } from 'react-router-dom'
import { ProductCardThumbnail } from './ProductCardThumbnail'
import { COLORS } from '../../utils/styleConstants'
import { DISCOUNT_PERCENTAGE_CARD_LIMIT } from '../../utils/constants'
import { LikeProductVars, LIKE_PRODUCT, DislikeProductData, DISLIKE_PRODUCT } from '../../apis/like'

export type Props = {
  width?: string
  lazyLoad?: boolean
  product: ProductCardType
  style?: React.CSSProperties
}

type StyledContainerProp = { width: string }

const StyledContainer = styled.div<StyledContainerProp>`
  width: ${(props) => props.width};
  flex: 0 0 auto;
`

const StyledContent = styled.div`
  line-height: 20px;
  margin-top: 4px;
  .price {
    font-weight: 700;
  }
`

const StyledDiscount = styled.div`
  font-size: 12px;
  padding: 1px 4px;
  background: ${COLORS.red};
  border-radius: 4px;
  color: white;
  position: absolute;
  left: 0;
  top: 0;
  z-index: 1;
`

const StyledLink = styled.a`
  color: inherit;
  display: block;
  position: relative;
  text-decoration: none;
`

export const ProductCard = (props: Props) => {
  const history = useHistory()
  let interval: any = null

  const { product, width = '50%', style, lazyLoad } = props
  const { id, price, name, thumbnailSrc, stockCount, discountRate } = product

  const [isLiked, setIsLiked] = useState(props.product.isLiked)

  const [likeProduct] = useMutation<{}, LikeProductVars>(LIKE_PRODUCT)
  const [dislikeProduct] = useMutation<DislikeProductData, LikeProductVars>(DISLIKE_PRODUCT)

  const toggleProductLike = async (e: React.MouseEvent) => {
    e.stopPropagation()

    const params = { variables: { productId: id } }
    if (isLiked) {
      await dislikeProduct(params)
    } else {
      await likeProduct(params)
    }

    setIsLiked(!isLiked)
  }

  const seperateClickEventHandler = (e: React.MouseEvent) => {
    e.preventDefault()

    if (interval) {
      toggleProductLike(e)
      clearTimeout(interval)
      return
    }

    interval = setTimeout(() => {
      history.push(`/product/${id}`)
      interval = null
    }, 200)
  }

  return (
    <StyledContainer className="product-card" width={width} style={style}>
      <StyledLink onClick={seperateClickEventHandler}>
        {discountRate >= DISCOUNT_PERCENTAGE_CARD_LIMIT && (
          <StyledDiscount className="discount">
            {`${discountRate}%`}
            <i className="icon">arrow_down</i>
          </StyledDiscount>
        )}
        <ProductCardThumbnail
          lazyLoad={lazyLoad}
          isLiked={isLiked}
          isSoldOut={stockCount === 0}
          thumbnailSrc={thumbnailSrc}
          toggleProductLike={toggleProductLike}
          seperateClickEventHandler={seperateClickEventHandler}
        />
        <StyledContent>
          <div className="product-name">{name}</div>
          <div className="price">{parseToLocalMoneyString(price)}Ïõê</div>
        </StyledContent>
      </StyledLink>
    </StyledContainer>
  )
}
