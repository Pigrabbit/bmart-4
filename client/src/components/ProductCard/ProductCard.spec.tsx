import React from 'react'
import { ProductCard } from './ProductCard'
import { mount } from 'enzyme'
import { ProductCardType } from '../../types/productCard'

const product: ProductCardType = {
  id: '5275',
  name: '테스트 상품명',
  price: 1000000,
  thumbnailSrc:
    'thumbnail7.coupangcdn.com/thumbnails/remote/492x492ex/image/retail/images/2019/08/09/15/2/267a0947-5985-43d7-8e65-5c0520d892e0.jpg',
}

describe('<ProductCard />', () => {
  it('matches snapshot', () => {
    const wrapper = mount(<ProductCard product={product} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders thumbnail, product name, price ', () => {
    const wrapper = mount(<ProductCard product={product} />)

    // check: props
    expect(wrapper.props().product).toBe(product)

    // check: product name
    const productNameElm = wrapper.find('div.product-name')
    expect(productNameElm.contains('테스트 상품명')).toBe(true)

    // check: price parsed to local string
    const priceElm = wrapper.find('div.price')
    expect(priceElm.text()).toBe('1,000,000원')

    // check: parsed thumbnail href
    const thumbnailElm = wrapper.find('img.thumbnail')
    expect(thumbnailElm.prop('src')).toBe(`http://${product.thumbnailSrc}`)
  })
})
