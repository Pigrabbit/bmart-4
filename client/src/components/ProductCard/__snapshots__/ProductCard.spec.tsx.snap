// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard /> match: snapshot 1`] = `ReactWrapper {}`;
